Background product image upload: show upload failure notice with an action to view product details

jaclync opened the pull request:
<!-- Remember about a good descriptive title. -->

Part of #7021 
<!-- Id number of the GitHub issue this PR addresses. -->

### Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->

When an image upload fails after the merchant leaves the product form, we want to show an in-app notice about the failure with an action to view product details. It was tricky to only notify the image upload updates when the merchant is **not** in the product form, since we currently show an error alert in the product form when an error occurs. We could also show an in-app notice in the product form, but we still want to customize the notice so that we don't show the action to view product details. I ended up adding two methods `stopEmittingStatusUpdates` and `startEmittingStatusUpdates` to pause the emission of a new `statusUpdates` publisher, when the product form view controller is shown/dismissed. If you have other ideas to implement this, please feel free to share!

Because the image upload error could come from any connected store, I decided to subscribe to the status updates in `MainTabBarController` to present the product (`ProductLoaderViewController`) from the products tab navigation controller after tapping the notice. I also chose to present the product instead of pushing the view controller so that the merchant can come back to any screen they were at (e.g. order or product form) if it's from the same store.

The changes still went over 500 diffs, most of them are on unit tests 🙇🏻‍♀️

### Testing instructions
<!-- Step by step testing instructions. When necessary break out individual scenarios that need testing, consider including a checklist for the reviewer to go through. -->

One way to test image upload failure is by updating the [media remote endpoint](https://github.com/woocommerce/woocommerce-ios/blob/0eae2a783db5f68577d613bf58b42cdda73e744f/Networking/Networking/Remote/MediaRemote.swift#L113) to an invalid one like adding some extra character at the end. You can also put `sleep(numberOfSeconds)` to simulate a longer response time.

Please note that there is a known issue (another subtask of #7021) where an error alert is shown every time you visit the product images screen after an upload failure.

#### Image upload failure outside of product form

- Launch the app
- In the product form, add one or more images
- Tap "Save" to save the product while the images are still being uploaded --> the product is saved remotely first, and there is no "Save" button anymore when the images are still being uploaded
- Leave the product form quickly, and feel free to do anything in the app including switching stores --> when an image fails to upload, an in-app notice (snackbar) should be shown with an action "View" (if you are viewing a modal when the failure occurs, the notice is shown after dismissing the modal)
- Tap on "View" to view product details --> it should switch to the original store if needed, and present the product form

#### Image upload failure within product form (production behavior)

- Launch the app
- In the product form, add one or more images
- Tap "Save" to save the product while the images are still being uploaded --> the product is saved remotely first, and there is no "Save" button anymore when the images are still being uploaded
- Stay in the product form and feel free to navigate within the form --> when an image fails to upload, an error alert should be shown like in production

### Screenshots
<!-- Include before and after images or gifs when appropriate. -->


https://user-images.githubusercontent.com/1945542/175632039-7b0d2ded-aa96-46d3-bebb-d5de6b0dbead.mov


---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

jaclync committed:
Add `statusUpdates`, `startEmittingStatusUpdates`, and `stopEmittingStatusUpdates` to support status updates and pause emission when the product form is present.

jaclync committed:
Stop and start observing status updates for the product when entering/leaving the product form.

jaclync committed:
Show product image upload failure notice with unit tests.

jaclync committed:
Add comments to public methods.

jaclync committed:
Update unit tests for product form invoking `startEmittingStatusUpdates` when it is dismissed or popped from the navigation stack.

selanthiraiyan commented on WooCommerce/Classes/ServiceLocator/ProductImageUploader.swift:
```diff
@@ -1,9 +1,21 @@
+
 /// Handles product image upload to support background image upload.
 protocol ProductImageUploaderProtocol {
+    /// Emits product image upload status updates (e.g. when the status changes or an error occurs).
```
> when the status changes

I believe this emits values only when at least one image failed to upload. 
https://github.com/woocommerce/woocommerce-ios/blob/fc47c8e0cd35dbb8bc06b7b44907c5d2915e2fdf/WooCommerce/Classes/ServiceLocator/ProductImageUploader.swift#L181
If the `error` is `nil` this will not be emitted. 

Should we update the comment and maybe rethink the variable name? 🤔

jaclync replied:
great question (and other similar questions below)! it was indeed quite confusing before, since we were still considering showing a notice for successful product updates before. Now that we decided not to (ref p91TBi-8Tm-p2#comment-10723), I renamed everything "status updates" to "errors" in https://github.com/woocommerce/woocommerce-ios/pull/7145/commits/30024d07e328e15f5ed981985aa0ea81031982f8

selanthiraiyan commented on WooCommerce/Classes/ServiceLocator/ProductImageUploader.swift:
```diff
@@ -124,3 +172,16 @@ final class ProductImageUploader: ProductImageUploaderProtocol {
 }
+
+private extension ProductImageUploader {
+    private func observeStatusUpdatesToUpdateProduct(key: ProductKey, actionHandler: ProductImageActionHandler) {
```
To me, the name `observeStatusUpdatesToUpdateProduct` conveys that we update the product by observing to status updates. But, we are just emitting the error events here. 
Am I missing something here?

jaclync replied:
This function was renamed to `observeStatusUpdatesForErrors` in https://github.com/woocommerce/woocommerce-ios/pull/7145/commits/30024d07e328e15f5ed981985aa0ea81031982f8 🙂

selanthiraiyan commented on WooCommerce/Classes/ViewRelated/MainTabBarController.swift:
```diff
@@ -553,3 +565,62 @@ private extension MainTabBarController {
+    }
+
+    func handleBackgroundImageUploadUpdate(_ update: ProductImageUploadUpdate) {
+        if update.error != nil {
```
What do you think about using a `guard` instead?

jaclync replied:
this check isn't necessary anymore after https://github.com/woocommerce/woocommerce-ios/commit/30024d07e328e15f5ed981985aa0ea81031982f8 😄

selanthiraiyan commented on WooCommerce/Classes/ViewRelated/MainTabBarController.swift:
```diff
@@ -553,3 +565,62 @@ private extension MainTabBarController {
+            let presenter = SwitchStoreNoticePresenter(siteID: update.siteID,
+                                                       noticePresenter: self.noticePresenter)
+            presenter.presentStoreSwitchedNoticeWhenSiteIsAvailable(configuration: .switchingStores)
+        }
```
Nice work covering the switching store scenario. 🙇 

Would it make sense to move the switch store logic to a separate function?

jaclync replied:
I feel like these lines of code are pretty straightforward and creating a separate function might not be easier to read 🤔 lemme know if you think otherwise!

selanthiraiyan commented on WooCommerce/Classes/ViewRelated/MainTabBarController.swift:
```diff
@@ -553,3 +565,62 @@ private extension MainTabBarController {
+private extension MainTabBarController {
+    enum Localization {
+        static let imageUploadFailureNoticeTitle =
+        NSLocalizedString("An image failed to upload",
```
Just thinking out loud here,
If more than one image failed to upload within a second, we will still show a single notice with the text `"An image failed to upload"`([as Notice presenter avoids showing duplicate notices](https://github.com/woocommerce/woocommerce-ios/blob/fc47c8e0cd35dbb8bc06b7b44907c5d2915e2fdf/WooCommerce/Classes/Tools/Notices/DefaultNoticePresenter.swift#L35)) which conveys that only one image upload failed. 🤔 

But, considering that we don't have a way to display the failed image uploads, this is out of scope. 😀

jaclync replied:
this is a very great point, and I agree that this is out of scope. When we show the failure state for each image in the UI in the future, this would be probably be less of an issue/confusion

selanthiraiyan commented on WooCommerce/Classes/ViewRelated/Products/Edit Product/ProductFormViewController.swift:
```diff
@@ -127,12 +130,26 @@ final class ProductFormViewController<ViewModel: ProductFormViewModelProtocol>:
+        /// A VC presented as the first VC in a navigation controller needs to check if the navigation controller is being dismissed.
+        /// A VC added to an existing navigation controller is dismissed when `isMovingFromParent` is `true`.
+        /// For any other scenario `isBeingDismissed` will do.
+        if isMovingFromParent || isBeingDismissed || navigationController?.isBeingDismissed == true {
```
Can we move this to a private computed property? 

Or would it make sense to move this to a `UIViewController` extension for reuse?  🤔

jaclync replied:
good suggestion, moved in https://github.com/woocommerce/woocommerce-ios/pull/7145/commits/48c98d5311a03ffb1280857c56148a3f6087a985. I borrowed this from [WPAuthenticator](https://github.com/wordpress-mobile/WordPressAuthenticator-iOS/blob/1adbca81f13756696e566a21d13c51365cc8075f/WordPressAuthenticator/Extensions/UIViewController%2BDismissal.swift#L5-L12) 😄

selanthiraiyan commented on WooCommerce/WooCommerceTests/ViewRelated/Products/Edit Product/ProductFormViewController+ProductImageUploaderTests.swift:
```diff
@@ -0,0 +1,101 @@
+
+@testable import WooCommerce
+
+final class ProductFormViewController_ProductImageUploaderTests: XCTestCase {
```
Nice tests covering all cases. ❤️🙇

selanthiraiyan reviewed (approved):
Great work. 🎉 

It works as expected. 

Thank you for adding unit tests. 💯 

**Question:**
Should we show an error notice when [the product update request](https://github.com/woocommerce/woocommerce-ios/blob/fc47c8e0cd35dbb8bc06b7b44907c5d2915e2fdf/WooCommerce/Classes/ViewRelated/Products/Media/ProductImagesSaver.swift#L61) fails?

jaclync committed:
Rename "status updates" to "errors" since we only need to emit errors.

jaclync committed:
Move `UIViewController` dismissal helper to an extension for potential reuse.

jaclync commented:
> **Question:**
> Should we show an error notice when [the product update request](https://github.com/woocommerce/woocommerce-ios/blob/fc47c8e0cd35dbb8bc06b7b44907c5d2915e2fdf/WooCommerce/Classes/ViewRelated/Products/Media/ProductImagesSaver.swift#L61) fails?

Great point! I added a new subtask to https://github.com/woocommerce/woocommerce-ios/issues/7021 and assigned myself to it 🙂 I'm going to work on this based on this branch.

@selanthiraiyan I updated the PR based on your suggestions, please feel free to take another look to double check the renaming!

selanthiraiyan commented on WooCommerce/WooCommerceTests/Mocks/MockProductImageUploader.swift:
```diff
@@ -1,9 +1,19 @@
+
     var replaceLocalIDWasCalled = false
     var saveProductImagesWhenNoneIsPendingUploadAnymoreWasCalled = false
+    var startEmittingStatusUpdatesWasCalled = false
```
Should we rename these variables as well? i.e. `startEmittingErrorsWasCalled`

selanthiraiyan commented on WooCommerce/WooCommerceTests/ViewRelated/Products/Edit Product/ProductFormViewController+ProductImageUploaderTests.swift:
```diff
@@ -0,0 +1,101 @@
+        super.tearDown()
+    }
+
+    func test_triggering_viewDidLoad_invokes_stopEmittingStatusUpdates() throws {
```
Should we rename these methods as well? i.e. `test_triggering_viewDidLoad_invokes_stopEmittingErrors`

selanthiraiyan commented on WooCommerce/WooCommerceTests/ViewRelated/Products/Media/ProductImageUploaderTests.swift:
```diff
@@ -193,4 +195,199 @@ final class ProductImageUploaderTests: XCTestCase {
+
+    // MARK: - Status Updates
+
+    func test_update_is_emitted_when_image_upload_fails() {
```
Perhaps `test_error_is_emitted_when_image_upload_fails`? And, can we follow the same for other test methods of this type as well?

selanthiraiyan commented on WooCommerce/WooCommerceTests/ViewRelated/Products/Media/ProductImageUploaderTests.swift:
```diff
@@ -193,4 +195,199 @@ final class ProductImageUploaderTests: XCTestCase {
                                                                      originalStatuses: []).productImageStatuses)
     }
+
+    // MARK: - Status Updates
```
Nit: `Error updates` maybe?

selanthiraiyan reviewed (approved):
Thanks for the updates, Jaclyn!

I left a few non-blocking comments.

:shipit:

jaclync commented:
@selanthiraiyan thanks for catching these naming issues, I'm going to address them in the followup PR https://github.com/woocommerce/woocommerce-ios/pull/7154 to avoid CI overhead.

jaclync commented:
Update: the naming issues were updated in https://github.com/woocommerce/woocommerce-ios/pull/7154/commits/46089d2f0f5c1d92abe5183b3bb4aae88a0892c3 🙂
